src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0004Tests.cs

```diff
@@ -496,5 +496,54 @@ await Verifier.CreateAnalyzer(code)
                 .WithDisabledDiagnostics("AZC0015")
                 .RunAsync();
         }
+
+        [Fact]
+        public async Task AZC0004NotProducedForConstructorOrPropertyOrOverride()
+        {
+            const string code = @"
+namespace RandomNamespace
+{
+    public class SomeClient
+    {
+        private string _id;
+        public virtual string Id => _id;
+
+        public override bool Equals(object obj) => base.Equals(obj);
+
+        protected SomeClient()
+        {
+        }
+
+        public SomeClient(string id)
+        {
+            _id = id;
+        }
+    }
+}";
+            await Verifier.CreateAnalyzer(code)
+                .RunAsync();
+        }
+
+        [Fact]
+        public async Task AZC0004NotProducedForGetSubClientMethod()
```

Comment:
Could we ever have a service method that ends with the `Client` suffix?  It feels more robust to me to check that the return type is a class in the library.

Reply:
We are using this rule to make sure service method is not returning arbitrary thing. Think about this method:
```C#
void ABC method(){}
```
How could you tell ABC is a sub client or an invalid return?

Reply:
I didn't have a specific definition in mind, but here are some thoughts - a subclient would have service methods on it in most cases?  So if a return type has service methods, we could have high confidence it is a subclient.

Do we ever have service methods that return types that are not: Response, Response\<T\>, Pageable\<T\>, Operation, Operation\<T\>?  This might be a simple way to differentiate between service methods and subclient methods.

Reply:
We should keep this in mind: we first get a method, then we start to check it. We check whether its input is valid and its return is valid. You could see that in my code, I also make sure the return type should be Response, Response<T>, Pageable<T>...etc...

In my example, it returns ABC and you say it is not a service method. Then what? We regard it as a subclient method and let it pass check? Doesn't it mean you don't even need to check it?

> So if a return type has service methods, we could have high confidence it is a subclient.

It is an infinite recursion, I'm afraid.